### ISSUE TITLE
[Feature Request] - Export HKWorkoutActivity sub-activities in workout JSON

### ISSUE THREAD

**Issue by sam234walker** [Lybron/health-auto-export#54]
**Is your feature request related to a problem? Please describe**.
When using custom/structured workouts on Apple Watch (e.g. Outdoor Run workout with custom intervals; warmup, effort & recovery blocks + cooldown), the Fitness app displays a full per-interval breakdown with individual statistics. However, when the same workout is exported via HAE's REST API automation, this interval structure is lost — only the parent workout summary is included. This makes it difficult to analyse performance differences between effort and recovery phases without manually correlating timestamps against the time-series heart rate data. Similar issue when adding a manual segment on an open session; the segmentation doesn't come through.

**Describe the solution you'd like**
Include the HKWorkout.workoutActivities array in the JSON v2 workout export. 

Each HKWorkoutActivity sub-activity contains:
- Boundaries: uuid, startDate, endDate, duration
- Also per-activity statistics via HKWorkoutActivity.statistics(for:): heart rate, distance, active energy, running speed, running power, ground contact time, vertical oscillation, stride length, step count (each with avg/min/max or sum). Preferably provide these within the proposed export addition. Some of those are currently covered under the Metrics auto export which I export separately, time match and tie back into analysis.
- Metadata: notably WOIntervalStepKeyPath (positional index into the custom workout template) and WOIntervalStepSuccessful

I've verified this data exists (with Claude) by examining Apple Health export XML — WorkoutActivity elements appear as children of Workout with full WorkoutStatistics and MetadataEntry children. 
This could be gated behind the existing "Include Workout Metrics" toggle to avoid increasing payload size for users who don't need it. The array would only be present on workouts that contain sub-activities.

**Describe alternatives you've considered**
- I currently reconstruct interval boundaries from time-series heart rate data by detecting pace/HR transitions — this works approximately but I tend to mix up the intervals and modalities so sometimes need adjust the detector logic as things change. It would be easier to assess the interval boundaries against the structural metadata (WOIntervalStepKeyPath) and/or utilising the pre-computed per-activity statistics from HealthKit.
- Using Apple's native Health export (Export All Health Data) — this does include WorkoutActivity elements in the XML, but it's a manual full dump with no filtering or automation capability.

**Additional context**
I use HAE's REST API automation to feed workout data to a self-hosted health analysis agent. The per-interval breakdown would significantly improve the quality of training analysis for structured sessions and simplify categorizing them.

**Device Information**
- Device: iPhone 15 Pro Max, Apple Watch Ultra 2
- OS: iOS 18.x, watchOS 11.6.1
- App Version: 9.0.8

**HAE User**
- Premium (Lifetime)

**Comment by Lybron** [comment]
Great outline of the feature request here, thanks for those details. 

Can definitely take a look into this.

**Comment by jbkomputer** [comment]
Chiming in as another user (via HAE sync to Mac) who'd use this feature. I run structured interval sessions on Apple Watch (warmup / [ efforts / recovery] x N / cooldown), and right now I reconstruct the rep boundaries after the fact by detecting shifts in power and cadence in the time-series. It works approximately, but Apple already stores the exact segmentation in workoutActivities, so getting it straight from the source would be a lot cleaner and more reliable.

If payload size is the worry, gating it behind the existing "Include Workout Metrics" toggle (as the OP suggested) would keep it opt-in for the rest of us.

Happy to beta-test with real structured-workout exports if that'd help move it along — thanks for considering it.

**Comment by jbkomputer** [comment]
Following up on my June comment with measurements, because I think the case for this is stronger than "it would be more convenient than reconstructing."

  I've now spent a full pass building a reconstructor and validating it against 33 reps across five structured sessions (4 × 200 m rep sessions + one 4 × 400 m), all
  Outdoor Run with programmed work blocks, exported via Sync to Mac. Two findings that bear on this request:

  **1. The block boundary is not a physical event, so it is not in the time series to be found.**

  A programmed workout advances into the work block the instant the recovery block completes. That boundary is a clock/distance event inside the watch. When the recovery
  is a jog — the normal case — the runner is already moving when it fires and simply accelerates through it. There is no step in speed, power, cadence or heart rate at
  that instant, because nothing physical happens at that instant.

  This is the part I'd underline: it isn't that reconstruction is *approximate*. For the rep onset, the information is **absent from the exported data**, not merely
  degraded in it. The only sessions where I can recover the onset are the ones where I happened to come to a near-stop at the line, and that's an artifact of my
  behaviour, not something the format guarantees.

  **2. The measured cost.**

  I segment reps by winding number on a closed track loop (the venue is a ~204 m oval, so a rep is one lap), which lets me measure onset placement geometrically:

  | session | how the recovery ended | onset scatter along the track |
  |---|---|---|
  | A | near-stop at the line | sd 4.0 m |
  | B | near-stop at the line | sd 4.6 m |
  | C | mixed | sd 5.6 m |
  | D | jogged, rolling | sd 9.0 m |
  | E | jogged, rolling | sd 32.9 m (178° of the loop) |

  On session E, two reps' reconstructed onsets sit on opposite sides of the track. Rep *distance* survives this (a full lap is a full lap wherever you start it), but the
  *split time* absorbs the whole error — roughly 1–7 s per rep, on reps that are ~48 s long. That's a 2–15% error on the single number the session exists to produce.

  For contrast, the quantities the export *does* carry are fine: ground distance per watch-measured 200 m came out at 202.5 m with sd 2.1 m over 30 reps. The data is
  good. It's specifically the segmentation that's missing, and it's the cheapest thing in the workout for HealthKit to hand over since Apple already has it in
  `workoutActivities`.

  Still happy to beta-test with real structured-workout exports whenever it's useful — I have five sessions on file with known ground truth (a measured track loop) that
  would make a decent validation set, and I can share the reconstructed-vs-actual comparison if that's helpful to you.

**Comment by RedelRedel** [comment]
Adding a +1 with a concrete use case and one measurement that I think sharpens the argument.

**Use case: 4x4 interval protocol.** Four work blocks with recovery in between, programmed as a structured workout on the Watch. The whole point of the analysis is the contrast between work and recovery — did the heart rate actually come down during the recovery, did the work blocks hold their pace, did the cadence hold up across repetitions. All of that needs the block boundaries. The parent workout summary averages work and recovery into numbers that describe neither.

**What reconstruction actually costs.** I built a reconstructor from the exported time series and validated it against known sessions. Two findings:

1. The reconstruction quality depends entirely on the export granularity of the metric you reconstruct from. Using the minute-aggregated `running_speed`, the same session read as four work blocks of 2.5–3.0 minutes. Using the per-second GPS route for the same session, it read as 3.6–3.8 minutes — which matches the programmed protocol. The minute aggregation smears the slow seconds at each block edge into the block itself. Anyone reconstructing without a GPS route is silently getting the wrong answer, and there is nothing in the export that warns them.

2. Even with per-second data, the boundary is a *plan* event, not a physical one. The workout advances into the work block the moment the timer says so; the runner accelerates over the following seconds. So the transition can be approximated but never recovered exactly — the exact instant only exists in `workoutActivities`.

**On payload size:** gating this behind the existing "Include Workout Metrics" toggle, as suggested upstream, would be entirely sufficient from my side. The segment list is tiny compared to the second-level metric series it would sit next to.

Happy to test a beta build against structured sessions if that helps.

**Comment by jbkomputer** [comment]
Confirming this works, with one regression and one gap, from the first structured session I ran after updating (HAE 10.0.x, JSON v2, All Workouts automation, Apple Watch Ultra 4, watchOS 27).

**Works.** A custom Outdoor Run (warm-in → 20:00 block → 6 × 200 m with jog recoveries → cooldown) exports `activities` with 15 entries, each with `start`/`end`/`activeDuration` and `metrics` for distance, speed, paceMile/paceKilometre, HR avg/min/max and activeEnergy (`origin: activityStatistics`). The 20-minute block comes through as one 1199 s activity; each rep as ~52 s / ~210 m. This is exactly what the issue asked for. Thank you.

**Regression: on that same workout every `segments[]` entry lost its distance and pace metrics.** All 18 auto-split segments (11 × 1 km, 7 × 1 mi) carry only `activeEnergy`, `heartRateAverage`, `heartRateMinimum`, `heartRateMaximum`:

```json
"metrics": {"activeEnergy": {...}, "heartRateAverage": {...}, "heartRateMinimum": {...}, "heartRateMaximum": {...}}
```

Every other segment-bearing workout in the same export folder (10 of 10 since 9/17, including a plain open-run warm-up recorded on the same watch 20 minutes earlier, same settings) carries the full set:

```json
"metrics": {"activeEnergy", "distance", "heartRateAverage", "heartRateMaximum", "heartRateMinimum", "paceKilometre", "paceMile", "speed"}
```

The one thing that separates the affected workout is `activities.length == 15` versus 1 everywhere else, so it looks like the segment-level distance overlap is skipped or fails once a workout has sub-activities. The data is present: summing the workout's own `walkingAndRunningDistance` samples over each segment window (excluding pause windows) reproduces the split lengths within 0.4%.

**Gap.** The activities carry no `metadata`, so `WOIntervalStepKeyPath` / `WOIntervalStepSuccessful` from the original request are not there; work vs recovery has to be inferred from pace. If that metadata is accessible, it would close the last piece.

Happy to send the full day file to support@ if useful.
